--- 
title: "Bombus haemolymph diapause study"
output: gene_level_differential_expression_diapause.html
authors:  Joe Colgan (joscolgan)
--- 
Scripts adapted from Colgan et al. (2019).  
Original script repository is here:
https://github.com/wurmlab/Bter_neonicotinoid_exposure_experiment

# Introduction
This script is for differential gene expression analysis of transcript quantification data.
Input files should be abundance files (.h5, .tsv, .json) generated by [Kallisto](https://pachterlab.github.io/kallisto/manual) 
for each sample and .txt file of genome transcripts 
and their corresponding gene ids (LOC numbers)/.

This script will import abundance files from Kallisto and use the R package [DESeq2](http://bioconductor.org/packages/release/bioc/html/DESeq.html) on count data to test for 
differential expression based on a model using negative binomial distribution.
This script produces a results table of genes that are significantly differentially expressed. 
This script uses the Wald test in DESeq2 to test for differential expression between treatments.
This can then be used for further analysis. 


```{r, message = FALSE}
# Load libraries; install from scratch if needed
libraries <- c("tximport", "readr", "DESeq2",
               "ggplot2", "gplots", "rhdf5",
               "plyr", "ggfortify", "biomaRt",
               "vsn", "RColorBrewer", "ReportingTools")
for (lib in libraries) {
    if (require(package = lib, character.only = TRUE)) {
        print("Successful")
    } else {
        print("Installing")
        source("https://bioconductor.org/biocLite.R")
        library(lib, character.only = TRUE )
    }
}
dir.create("results")
```


## Step One: Input results from kallisto

We use raw, non-normalised estimated counts.
```{r} 
## Set paths to folders containing output files from kallisto
paths <- list()
paths$sample_batch_info <- "./sample_information.txt"
paths$kallisto_output   <- "input/"

# Set relative paths:
paths$kallisto_files_relative <- grep(x = list.files(paths$kallisto_output,
                                                     recursive = TRUE),
                                      pattern = ".h5",
                                      value   = TRUE)
paths$kallisto_files <- file.path(paths$kallisto_output,
                                  paths$kallisto_files_relative)

# Automatically extract file names from kallisto output for colnames
names(paths$kallisto_files) <- gsub(paths$kallisto_files_relative,
                                    pattern = "/.*",
                                    replacement = "")

for (filenumber in 1:length(paths$kallisto_files)) {
  current_name <- names(paths$kallisto_files)[filenumber]
  current_file <- paths$kallisto_files[filenumber]
  if (FALSE == grepl(pattern = current_name, x = current_file)) {
    kill("we have a problem - names and filenames dont match up")
  }
}
```

## Step Two: Put sample names and treatments into a samples dataframe for DESeqDataSetFromTximport

```{r, message = FALSE}
## Extract sample names and put into df for tximport
samples     <- data.frame(treatment = names(paths$kallisto_files))

## Read in sample information for room and batch:
samples_information <- read.table(file = paths$sample_batch_info,
                                  header = FALSE,
                                  col.names = c("sample_name",
                                                "treatment",
                                                "platform"),
                                  row.names = 1)

## Make sure 'CON' is the reference:
samples_information$treatment <- relevel(x = samples_information$treatment,
                                         ref = "Diapause")

# sanity check
if (FALSE == all(samples$treatment %in% rownames(samples_information))) {
    kill("we have a problem - sample names from files and
         from batch descriptor file  dont match up")
}
samples_information
```

## Step Three: Import kallisto quantification files using Tximport

We used kallisto to obtain estimated counts for each transcript in each condition. 
We load these raw counts into R for DEseq using tximport. 
Because we do a gene-level analysis first, transcript abundances are summarised per gene during import.

```{r}
## Read in genes and corresponding RNA ids:
transcript_to_gene <- read.table(file = "./rna_and_corresponding_genes.txt",
                                 header = FALSE)

# Use tximport on kallisto files 
## Counts are estimated counts from kallisto:
txi_counts <- tximport(paths$kallisto_files,
                       type    = "kallisto",
                       tx2gene = transcript_to_gene,
                       countsFromAbundance = "no")

# Save object for use in other scripts
dir.create(path = "results/count_estimates/",
           recursive = TRUE)
save(txi_counts, file = "results/count_estimates/txi_count_estimates.Rdata")

## Convert counts to dataframe:
txi_counts_df <- as.data.frame(txi_counts$counts)

# Write gene-level estimated counts to file:
write.table(txi_counts_df,
            file = "./results/raw_gene_level_counts.txt",
            row.names = TRUE,
            sep = "\t",
            quote = FALSE)
```

### Step Four: Run DESeq2 for differential expression analysis:

For identifying differentially expressed genes, a DESeq object is generated containing:
    a. sample names
    b. treatments
    
Within the present analysis, four questions can be immediately asked:
    1. What genes are differentially expressed within clothianidin-exposed workers in comparison to control?
    2. What genes are differentially expressed within imidacloprid-exposed workers in comparison to control?

```{r}
## Construct DESeq data set from txi object and sample information:
deseq_txi <- DESeqDataSetFromTximport(txi     = txi_counts,
                                      colData = samples_information,
                                      design  = ~platform + treatment)
```

# Run DESeq2 (Wald test) for differential expression analysis:
  Which genes are differentially expressed between each treatment and control?
  Running a Wald t-test (default test within DESeq2)
  
```{r}
## Which genes are differentially expressed within each treatment for
## each caste in comparison to control?
## Perform a LRT test to identify genes differentially expressed across timepoints:
deseq_object_lrt  <- DESeq(deseq_txi,
                           test = "LRT",
                           reduced = ~platform,
                           betaPrior = FALSE)

## Explore results:
mated_vs_diapause_lrt <- results(deseq_object_lrt,
                                 contrast = c("treatment",
                                              "Mated_prediapause",
                                              "Diapause"))
post_vs_diapause_lrt <- results(deseq_object_lrt,
                                contrast = c("treatment",
                                             "Founder_postdiapause",
                                             "Diapause"))

## Perform a pairwise comparison between treatments:
deseq_object_wald  <- DESeq(deseq_txi,
                            test = "Wald",
                            betaPrior = TRUE)

## Explore results:
mated_vs_diapause_wt <- results(deseq_object_wald,
                                 contrast = c("treatment",
                                              "Mated_prediapause",
                                              "Diapause"))
post_vs_diapause_wt <- results(deseq_object_wald,
                                contrast = c("treatment",
                                             "Founder_postdiapause",
                                             "Diapause"))

## Subset genes identified using both approaches:
mated_vs_diapause_lrt_sig <- subset(mated_vs_diapause_lrt,
                                    padj < 0.05)
nrow(mated_vs_diapause_lrt_sig)

mated_vs_diapause_wt_sig <- subset(mated_vs_diapause_wt,
                                    padj < 0.05)
nrow(mated_vs_diapause_wt_sig)

## Subset:
overlap_sig_pre <- subset(mated_vs_diapause_lrt_sig,
                      rownames(mated_vs_diapause_lrt_sig) %in%
                              rownames(mated_vs_diapause_wt_sig))

## Convert to dataframe:
overlap_sig_pre_df <- as.data.frame(overlap_sig_pre)

# Order by p-values 
overlap_sig_pre_df <- overlap_sig_pre_df[order(overlap_sig_pre_df$pvalue), ]

## Write to output:
write.table(overlap_sig_pre_df,
            file = "results/de_genes_pre_vs_diapause.sig.txt",
            row.names = TRUE,
            sep = "\t",
            quote = FALSE)

## Subset genes identified using both approaches:
post_vs_diapause_lrt_sig <- subset(post_vs_diapause_lrt,
                                    padj < 0.05)
nrow(post_vs_diapause_lrt_sig)

post_vs_diapause_wt_sig <- subset(post_vs_diapause_wt,
                                    padj < 0.05)
nrow(post_vs_diapause_wt_sig)

## Subset:
overlap_sig_post <- subset(post_vs_diapause_lrt_sig,
                      rownames(post_vs_diapause_lrt_sig) %in%
                              rownames(post_vs_diapause_wt_sig))

## Convert to dataframe:
overlap_sig_post_df <- as.data.frame(overlap_sig_post)

# Order by p-values 
overlap_sig_post_df <- overlap_sig_post_df[order(overlap_sig_post_df$pvalue), ]

## Write to output:
write.table(overlap_sig_post_df,
            file = "results/de_genes_post_vs_diapause.sig.txt",
            row.names = TRUE,
            sep = "\t",
            quote = FALSE)


```

Generate plot using principal components calculated for transcriptomes from all queens:

```{r}
## Extracting transformed values from DESeq object:
vsd <- vst(deseq_object_wald,
           blind = FALSE)
## Basic plot:
plotPCA(vsd, intgroup = c("treatment"))
## Extract first and second principal components:
pca_data    <- plotPCA(vsd,
                       intgroup = c("treatment"),
                       returnData = TRUE)
## Extract percentages for first two principal components:
percentvar <- round(100 * attr(pca_data, "percentVar"))
## Plot PCA:
ggplot(pca_data, aes(PC1, PC2,
                     color = treatment,
                     shape = treatment)) +
        geom_point(aes(fill = treatment),
                   colour = "black", pch = 21, size = 5) +
        xlab(paste0("PC1: ", percentvar[1], "% variance")) +
        ylab(paste0("PC2: ", percentvar[2], "% variance")) +
        coord_fixed() +
        scale_color_manual(values = c("blue", "grey", "black")) +
        theme_bw() +
        theme(axis.title = element_text(family = "Arial",
                                        color = "#666666",
                                        face = "bold",
                                        size = 16)) +
        theme(legend.position = "top") +
        theme(legend.text = element_text(colour = "black",
                                         size = 14,
                                         face = "bold"),
              legend.title = element_text(colour = "black",
                                          size = 14,
                                          face = "bold"))
## Save file:
ggsave(file = "pca_output_diapause_samples.png",
       height = 10,
       width = 10)
```

Check style guide as a final step

```{r, message = FALSE}
## Run lintr:
library(lintr)
lintr::lint("./gene_level_differential_expression_diapause.Rmd")
```
