---
title: "Bombus haemolymph diapause study"
output: plot_phylogenetic_tree.html
authors: Joe Colgan (joscolgan)
---

## Introduction:  
The purpose of this script is to take a newick-aligned dataset of protein sequences and plot the phylogeny.  
The input was generated using [www.phylogeny.fr](http://www.phylogeny.fr/) with the phylogenetic tree visualised using [ggtree](https://bioconductor.org/packages/release/bioc/html/ggtree.html).   
The script outputs a phylogenetic tree in PNG and PDF format.

Load libraries:

```{r, message = FALSE}
## Load libraries:
libraries <- c("ggplot2",
               "ggtree",
               "lintr")
for (lib in libraries) {
        if (require(package = lib, character.only = TRUE)) {
                print("Successful")
        } else {
                print("Installing")
                source("https://bioconductor.org/biocLite.R")
                biocLite(pkgs = lib)
                library(lib, character.only = TRUE )
        }
}
```

Read in data:

```{r, message = FALSE}
## Read newick alignment file in and store as variable:
tree <- read.tree("./data/hahp_phylogeny_newick_format_new.txt")

## Annotation tip label for plotting species name:
groupInfo <- split(tree$tip.label,
                   gsub("_\\w+", "", tree$tip.label))

## Plot tree:
tree <- groupOTU(tree, groupInfo)
ggtree(tree, aes(color = group),
       layout = "circular") +
        geom_tiplab(size = 3.2,
                    aes(angle = angle)) +
        geom_treescale(fontsize = 5)

## Create output directory:
dir.create(path = "./results/png/",
           recursive = TRUE)

## Define output:
output_png <- "results/png/phylo_plot.png"

## Save output:
ggsave(output_png,
       height = 12,
       width = 16)

## Create output directory:
dir.create(path = "./results/pdf/",
           recursive = TRUE)

## Define output:
output_pdf <- "results/pdf/phylo_plot.pdf"

## Save output:
ggsave(output_pdf,
       dpi = 600,
       height = 11,
       width = 16)

```
Run lintr

```{r, message = TRUE}
lintr::lint(file = "./plot_phylogenetic_tree.Rmd")
```

